### PR TITLE
[SPARK-22333][SQL]timeFunctionCall(CURRENT_DATE, CURRENT_TIMESTAMP) has conflicts with columnReference

### PR DESCRIPTION
**Pull request by DonnyZone**
## What changes were proposed in this pull request?
https://issues.apache.org/jira/browse/SPARK-22333

In current version, users can use CURRENT_DATE() and CURRENT_TIMESTAMP() without specifying braces.
However, when a table has columns named as "current_date" or "current_timestamp", it will still be parsed as function call.

There are many such cases in our production cluster. We get the wrong answer due to this inappropriate behevior. In general, ColumnReference should get higher priority than timeFunctionCall.

## How was this patch tested?
unit test
manul test

### CODE REVIEW

**Comment by DonnyZone** [comment]
ping @hvanhovell @gatorsmile

**Comment by DonnyZone** [comment]
ping @gatorsmile @hvanhovell @cloud-fan

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 1]
```diff
@@ -585,6 +584,7 @@ primaryExpression
     | identifier                                                                               #columnReference
     | base=primaryExpression '.' fieldName=identifier                                          #dereference
     | '(' expression ')'                                                                       #parenthesizedExpression
+    | name=(CURRENT_DATE | CURRENT_TIMESTAMP)                                                  #timeFunctionCall
```
Comment (hvanhovell) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Won't this break every use of `CURRENT_DATE`/`CURRENT_TIMESTAMP`? They will now be interpreted as an identifier.

**Comment by hvanhovell** [comment]
ok to test

**Comment by DonnyZone** [comment]
@hvanhovell Yes! I made something wrong. The `timeFunctionCall` has conflicts with `columnReference`. This fix will break every use of CURRENT_DATE/CURRENT_TIMESTAMP.

For [SPARK-16836](https://github.com/apache/spark/pull/14442), 
I think this feature should be implemented in analysis phase rather than in parser phase. When there is no such columns, they can be transformed as functions. Another approach is to define a configuration.

**Comment by gatorsmile** [comment]
@DonnyZone Analyzer is the best place to fix the issue.

**Comment by DonnyZone** [comment]
@gatorsmile Thank for your advice, I will work on it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -139,6 +139,7 @@ class Analyzer(
       ExtractGenerator ::
       ResolveGenerate ::
       ResolveFunctions ::
+      ResolveLiteralFunctions ::
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: The order matters. It assumes `ResolveReferences` should be run before this rule. However, `ResolveReferences` might need multiple passes to resolve all the references. Thus, how about moving the logics into `ResolveReferences `? If the attributes are not resolvable, we try to see whether it is a function literal?

Comment (DonnyZone) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Agree! I will refactor it.

**Comment by gatorsmile** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -783,6 +783,25 @@ class Analyzer(
...
+     */
+    private def resolveAsLiteralFunctions(nameParts: Seq[String]): Option[NamedExpression] = {
+      if (nameParts.length != 1) {
+        return None
+      }
+      // support CURRENT_DATE and CURRENT_TIMESTAMP
+      val literalFunctions = Seq(CurrentDate(), CurrentTimestamp())
+      val name = nameParts.head
+      val func = literalFunctions.find(e => resolver(e.prettyName, name))
+      if (func.isDefined) {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Just map over the `func` option.

Comment (DonnyZone) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Thanks, I will refactor it.

**Comment by DonnyZone** [comment]
@gatorsmile @gatorsmile 
There are still two issues need to be figured out.
(1)It will be complicated to determine whether a literal function should be resolved as Expression or NamedExpression.
Current fix just resolves them as NamedExpressions (i.e., Alias). 
However, this leads to different schema in some cases, for example, the end-to-end test sql.
`select current_date = current_date()`
The output schema will be
`struct<(current_date() AS ’current_date()‘ = current_date()):boolean>`
(2)Shall we also support the feature in ResolveMissingReference rule?
e.g., `select id from table order by current_date`
The same logic in different rules brings redundant code.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 4]
```diff
@@ -844,7 +863,12 @@ class Analyzer(
         q.transformExpressionsUp  {
           case u @ UnresolvedAttribute(nameParts) =>
             // Leave unchanged if resolution fails. Hopefully will be resolved next round.
-            val result = withPosition(u) { q.resolveChildren(nameParts, resolver).getOrElse(u) }
+            val result =
+              withPosition(u) {
+                q.resolveChildren(nameParts, resolver).getOrElse {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: You can also use `orElse`:
`q.resolveChildren(nameParts, resolver).orElse(resolveAsLiteralFunctions(nameParts)).getOrElse(u)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 5]
```diff
@@ -783,6 +783,37 @@ class Analyzer(
...
+        case p @ Project(projList, _) if (projList.contains(attribute)) =>
+          true
+        case _ =>
+          false
+      }
+      val wrapper: Expression => Expression = if (isNamedExpression) {
+        f => Alias(f, toPrettySQL(f))()
+      } else {
+        f => f
+      }
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]:
```Scala
      if (nameParts.length != 1) return None
      val isNamedExpression = plan match {
        case Aggregate(_, aggs, _) => aggs.contains(attribute)
        case Project(projList, _) => projList.contains(attribute)
        case Window(windowExpressions, _, _, _) => windowExpressions.contains(attribute)
        case _ => false
      }
      val wrapper: Expression => Expression =
        if (isNamedExpression) f => Alias(f, usePrettyExpression(f).sql)() else identity
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 6]
```diff
@@ -783,6 +783,37 @@ class Analyzer(
...
+        plan: LogicalPlan): Option[Expression] = {
+      if (nameParts.length != 1) {
+        return None
+      }
+      val isNamedExpression = plan match {
+        case a @ Aggregate(_, aggs, _) if (aggs.contains(attribute)) =>
+          true
+        case p @ Project(projList, _) if (projList.contains(attribute)) =>
+          true
+        case _ =>
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Miss `Windows`

File: sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala [hunk 7]
```diff
@@ -56,6 +56,24 @@ class DateFunctionsSuite extends QueryTest with SharedSQLContext {
...
+
+    val df1 = Seq((1, 2), (2, 3)).toDF("a", "b")
+    df1.createOrReplaceTempView("ttf1")
+    withTempView("ttf1") {
+      checkAnswer(
+        sql("SELECT current_date = current_date(), current_timestamp = current_timestamp(), " +
+          "a, b FROM ttf1"),
+        Seq(Row(true, true, 1, 2), Row(true, true, 2, 3)))
+    }
+  }
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala]: Move these to `datetime.sql`?

File: sql/core/src/test/resources/sql-tests/inputs/datetime.sql [hunk 8]
```diff
@@ -8,3 +8,18 @@ select to_date(null), to_date('2016-12-31'), to_date('2016-12-31', 'yyyy-MM-dd')
...
+  as ttf1(current_date, current_timestamp);
+  
+select current_date, current_timestamp from ttf1
+
+create temporary view ttf2 as select * from values
+  (1, 2),
+  (2, 3),
+  as ttf2(a, b);
+  
+select current_date = current_date(), current_timestamp = current_timestamp(), a, b from ttf2
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/datetime.sql]: Add an empty line

Comment (DonnyZone) [reply sql/core/src/test/resources/sql-tests/inputs/datetime.sql]: Fixed.

**Comment by gatorsmile** [comment]
LGTM and cc @hvanhovell for a final sign-off. 

BTW, I might merge https://github.com/apache/spark/pull/19585 before this PR. Sorry for the conflicts.

**Comment by DonnyZone** [comment]
@gatorsmile 
It seems that we should also support this logic in `resolveExpressions` for Sort plan.
`select a from t order by current_date`
Therefore, I think current `resolveAsLiteralFunctions` can be moved out from `ResolveReference` rule to be  a common function `resolveLiteralFunctions`

**Comment by DonnyZone** [comment]
OK, I will close this PR after review and submit a new one, after merging https://github.com/apache/spark/pull/19585

**Comment by gatorsmile** [comment]
You do not need to close it and just need to resolve the conflicts. If you are not familiar with it, I can wait and merge this PR first.

**Comment by gatorsmile** [comment]
`select a from t order by current_date` is not working?

**Comment by DonnyZone** [comment]
Yes, ordering in `Sort(ordering, global, child)` is resolved in `resolveExpression`

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Comment by gatorsmile** [comment]
Could you submit a backport PR to 2.2?

**Comment by DonnyZone** [comment]
Sure, I will submit it later.

**Pull request closed without merging**